File: platform/typescript/kernel-product-contracts/src/route/ProductRouteContract.ts
/**
 * K-001: Kernel product route contract schema.
 * Product route metadata is validated once and projected into UI manifests,
 * backend entitlement payloads, and route capability summaries.
 */

import { z } from "zod";

export const RouteStabilityValues = ["stable", "preview", "blocked", "hidden"] as const;
export const RouteGroupValues = [
  "care",
  "governance",
  "clinical",
  "administrative",
  "profile",
  "dashboard",
  "system",
  "emergency",
  "provider",
  "caregiver",
  "fchv",
] as const;

export const RouteHttpMethodValues = ["GET", "POST", "PUT", "DELETE", "PATCH"] as const;
export const RouteActionVisibilityValues = ["public", "authenticated", "role-restricted"] as const;

export type RouteStability = (typeof RouteStabilityValues)[number];
export type RouteGroup = (typeof RouteGroupValues)[number];

const RouteStabilitySchema = z.enum(RouteStabilityValues);
const RouteGroupSchema = z.enum(RouteGroupValues);

export const RouteMetadataSchema = z
  .object({
    apiEndpoint: z.string().trim().min(1).optional(),
    policyId: z.string().trim().min(1).optional(),
    testId: z.string().trim().min(1).optional(),
    featureFlag: z.string().trim().min(1).optional(),
    introducedAt: z.string().trim().min(1).optional(),
  })
  .strict();

export const RouteActionSchema = z
  .object({
    id: z.string().trim().min(1),
    label: z.string().trim().min(1),
    endpoint: z.string().trim().min(1),
    method: z.enum(RouteHttpMethodValues),
    policyId: z.string().trim().min(1).optional(),
    idempotent: z.boolean().optional(),
    confirmationRequired: z.boolean().optional(),
    visibility: z.enum(RouteActionVisibilityValues).optional(),
  })
  .strict();

// Allow string references for actions/cards (product-specific action IDs)
export const RouteActionRefSchema = z.string().trim().min(1);
export const RouteCardRefSchema = z.string().trim().min(1);

export const RouteCardSchema = z
  .object({
    id: z.string().trim().min(1),
    title: z.string().trim().min(1),
    description: z.string().trim().min(1),
    icon: z.string().trim().min(1).optional(),
    badge: z.string().trim().min(1).optional(),
  })
  .strict();

export const ProductRouteSchema = z
  .object({
    path: z.string().trim().min(1).startsWith("/"),
    label: z.string().trim().min(1),
    description: z.string().trim().min(1),
    group: RouteGroupSchema,
    minimumRole: z.string().trim().min(1),
    personas: z.array(z.string().trim().min(1)).optional(),
    tiers: z.array(z.string().trim().min(1)).optional(),
    actions: z.array(z.union([RouteActionSchema, RouteActionRefSchema])).optional(),
    cards: z.array(z.union([RouteCardSchema, RouteCardRefSchema])).optional(),
    stability: RouteStabilitySchema,
    featureFlag: z.boolean().optional(),
    metadata: RouteMetadataSchema.optional(),
    // Product-specific extensions
    surface: z.array(z.enum(["web", "mobile", "backend", "hidden"])).optional(),
    i18nKey: z.string().trim().min(1).optional(),
    descriptionI18nKey: z.string().trim().min(1).optional(),
    routeType: z.enum(["page", "detail", "action", "system"]).optional(),
    visibilityReason: z.string().trim().min(1).optional(),
    emergencyAction: z.boolean().optional(),
    // Allow direct placement of metadata fields for backward compatibility
    apiEndpoint: z.string().trim().min(1).optional(),
    policyId: z.string().trim().min(1).optional(),
    testId: z.string().trim().min(1).optional(),
  })
  .strict()
  .superRefine((route, context) => {
    if (route.stability !== "stable") {
      return;
    }

    // Check metadata object first, then direct fields
    const apiEndpoint = route.metadata?.apiEndpoint ?? route.apiEndpoint;
    const policyId = route.metadata?.policyId ?? route.policyId;
    const testId = route.metadata?.testId ?? route.testId;

    const missingMetadata = [
      apiEndpoint ? undefined : "apiEndpoint",
      policyId ? undefined : "policyId",
      testId ? undefined : "testId",
    ].filter((value): value is string => value !== undefined);

    if (missingMetadata.length > 0) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["metadata"],
        message: `Stable route ${route.path} is missing ${missingMetadata.join(", ")}`,
      });
    }
  });

export const ProductRouteContractSchema = z
  .object({
    version: z.string().trim().min(1).optional(),
    schemaVersion: z.string().trim().min(1).optional(),
    product: z.string().trim().min(1).optional(),
    roleOrder: z.record(z.string().trim().min(1), z.number().int().nonnegative()),
    routes: z.array(ProductRouteSchema).min(1),
  })
  .strict()
  .superRefine((contract, context) => {
    if (!contract.version && !contract.schemaVersion) {
      context.addIssue({
        code: z.ZodIssueCode.custom,
        path: ["version"],
        message: "Route contract must declare version or schemaVersion",
      });
    }

    const seenPaths = new Set<string>();

    contract.routes.forEach((route, index) => {
      if (seenPaths.has(route.path)) {
        context.addIssue({
          code: z.ZodIssueCode.custom,
          path: ["routes", index, "path"],
          message: `Duplicate route path ${route.path}`,
        });
      }
      seenPaths.add(route.path);

      if (contract.roleOrder[route.minimumRole] === undefined) {
        context.addIssue({
          code: z.ZodIssueCode.custom,
          path: ["routes", index, "minimumRole"],
          message: `Route ${route.path} references unknown minimumRole ${route.minimumRole}`,
        });
      }
    });
  })
  .transform(contract => ({
    ...contract,
    version: contract.version ?? contract.schemaVersion ?? '',
  }));

export type RouteAction = z.infer<typeof RouteActionSchema>;
export type RouteCard = z.infer<typeof RouteCardSchema>;
export type RouteMetadata = z.infer<typeof RouteMetadataSchema>;
export type ProductRoute = z.infer<typeof ProductRouteSchema>;
export type ProductRouteContract = z.infer<typeof ProductRouteContractSchema>;

export type ProductRouteCapability = {
  path: string;
  stability: RouteStability;
  directLinkAllowed: boolean;
  discoverable: boolean;
  minimumRole: string;
  featureFlag?: boolean;
  apiEndpoint?: string;
  policyId?: string;
  testId?: string;
};

export const ProductRouteCapabilitySchema = z
  .object({
    path: z.string().trim().min(1).startsWith("/"),
    stability: RouteStabilitySchema,
    directLinkAllowed: z.boolean(),
    discoverable: z.boolean(),
    minimumRole: z.string().trim().min(1),
    featureFlag: z.boolean().optional(),
    apiEndpoint: z.string().trim().min(1).optional(),
    policyId: z.string().trim().min(1).optional(),
    testId: z.string().trim().min(1).optional(),
  })
  .strict();

export function isRouteStability(value: unknown): value is RouteStability {
  return RouteStabilitySchema.safeParse(value).success;
}

export function isRouteGroup(value: unknown): value is RouteGroup {
  return RouteGroupSchema.safeParse(value).success;
}

export function parseProductRouteContract(contract: unknown): ProductRouteContract {
  return ProductRouteContractSchema.parse(contract);
}

export function validateProductRouteContract(contract: unknown): contract is ProductRouteContract {
  return ProductRouteContractSchema.safeParse(contract).success;
}

export function validateProductRouteCapability(value: unknown): value is ProductRouteCapability {
  return ProductRouteCapabilitySchema.safeParse(value).success;
}

File: platform/java/contracts/src/main/java/com/ghatana/contracts/route/ProductRouteContract.java
package com.ghatana.contracts.route;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Shared route contract model for product routes.
 *
 * <p>This model mirrors the TypeScript ProductRouteContract schema in
 * kernel-product-contracts, providing a shared contract definition for
 * both TypeScript and Java ecosystems.
 *
 * @doc.type record
 * @doc.purpose Shared route contract model for TS and Java
 * @doc.layer platform
 * @doc.pattern Data Transfer Object
 */
public record ProductRouteContract(
    String version,
    String schemaVersion,
    String product,
    Map<String, Integer> roleOrder,
    List<ProductRoute> routes
) {
    /**
     * Product route definition.
     */
    public record ProductRoute(
        String path,
        String label,
        String description,
        String group,
        String minimumRole,
        List<String> personas,
        List<String> tiers,
        List<Object> actions,
        List<Object> cards,
        String stability,
        Boolean featureFlag,
        RouteMetadata metadata,
        List<String> surface,
        String i18nKey,
        String descriptionI18nKey,
        String routeType,
        String visibilityReason,
        Boolean emergencyAction,
        String apiContractId,
        String dtoSchemaId,
        List<String> pluginDependencies,
        String auditRequirement,
        String phiSensitivity,
        String cachePolicy,
        String offlinePolicy
    ) {
        /**
         * Route metadata.
         */
        public record RouteMetadata(
            String apiEndpoint,
            String policyId,
            String testId,
            String featureFlag,
            String introducedAt
        ) {}
    }

    /**
     * Route action definition.
     */
    public record RouteAction(
        String id,
        String label,
        String endpoint,
        String method,
        String policyId,
        Boolean idempotent,
        Boolean confirmationRequired,
        String visibility
    ) {}

    /**
     * Route card definition.
     */
    public record RouteCard(
        String id,
        String title,
        String description,
        String icon,
        String badge
    ) {}

    /**
     * Route stability values.
     */
    public enum RouteStability {
        STABLE,
        PREVIEW,
        BLOCKED,
        HIDDEN,
        DEFERRED,
        REMOVED
    }

    /**
     * Route group values.
     */
    public enum RouteGroup {
        CARE,
        GOVERNANCE,
        CLINICAL,
        ADMINISTRATIVE,
        PROFILE,
        DASHBOARD,
        SYSTEM,
        EMERGENCY,
        PROVIDER,
        CAREGIVER,
        FCHV
    }

    /**
     * HTTP method values.
     */
    public enum HttpMethod {
        GET,
        POST,
        PUT,
        DELETE,
        PATCH
    }

    /**
     * Audit requirement values.
     */
    public enum AuditRequirement {
        NONE,
        STANDARD,
        PHI_ACCESS,
        PHI_WRITE,
        EMERGENCY_BREAK_GLASS,
        ADMIN_REVIEW
    }

    /**
     * PHI sensitivity values.
     */
    public enum PhiSensitivity {
        NONE,
        PII,
        PHI,
        RESTRICTED_PHI,
        EMERGENCY_PHI
    }

    /**
     * Cache policy values.
     */
    public enum CachePolicy {
        NO_STORE,
        PRIVATE_SESSION,
        SHORT_LIVED,
        OFFLINE_ENCRYPTED
    }

    /**
     * Offline policy values.
     */
    public enum OfflinePolicy {
        ONLINE_ONLY,
        METADATA_ONLY,
        ENCRYPTED_TTL,
        EMERGENCY_UNAVAILABLE
    }
}
